### PR TITLE
In-memory Engine: refactor from region based to range based

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
refactor from region based to range based
```

This PR refactors region based in-memory engine to range based in-memory engine to decouple in-memory engine with raftstore as much as possible. It means in-memory engine is ignorant about admin commands such as batch split and merge.
This PR provides a `RangeManager` which manages the ranges cached.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/engine_traits/src/memory_engine.rs [hunk 1]
```diff
@@ -1,19 +1,110 @@
...
-use std::fmt::Debug;
+use std::{cmp, fmt::Debug};
+
+use keys::{enc_end_key, enc_start_key};
+use kvproto::metapb;
 
 use crate::{Iterable, Snapshot, WriteBatchExt};
 
-/// RegionCacheEngine works as a region cache caching some regions (in Memory or
+/// RangeCaheEngine works as a range cache caching some ranges (in Memory or
```
Comment (Connor1996) [inline components/engine_traits/src/memory_engine.rs:10]:
```suggestion
/// RangeCacheEngine works as a range cache caching some ranges (in Memory or
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 2]
```diff
@@ -813,22 +874,23 @@ mod tests {
 
     #[test]
     fn test_get_value() {
-        let engine = RegionCacheMemoryEngine::default();
-        engine.new_region(1);
+        let engine = RangeCacheMemoryEngine::new(Arc::default());
+        let range = CacheRange::new(b"k000".to_vec(), b"k100".to_vec());
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:878]: are we testing the edge condition where one of the cached ranges is open one end or the other? e.g., from first data key, to the end?

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+impl RangeMeta {
+    pub(crate) fn range_snapshot_list(&self) -> &BTreeMap<CacheRange, SnapshotList> {
+        &self.range_snapshot_list
+    }
+
+    pub(crate) fn merge_meta(&mut self, mut other: RangeMeta) {
+        self.range_snapshot_list
+            .append(&mut other.range_snapshot_list);
+        self.ranges_evcited.append(&mut other.ranges_evcited);
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/range_manager.rs:25]:
```suggestion
        self.ranges_evicted.append(&mut other.ranges_evicted);
```

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            return false;
+        };
+        let meta = self.ranges.get(range_key).unwrap();
+
+        if read_ts <= meta.safe_ts {
+            // todo(SpadeA): add metrics for it
+            return false;
+        }
+
+        if meta.ranges_evcited.iter().any(|r| r.overlaps(range)) {
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/range_manager.rs:120]:
```suggestion
        if meta.ranges_evicted.iter().any(|r| r.overlaps(range)) {
```

File: components/region_cache_memory_engine/src/write_batch.rs [hunk 5]
```diff
@@ -2,16 +2,16 @@ use bytes::Bytes;
 use engine_traits::{Mutable, Result, WriteBatch, WriteBatchExt, WriteOptions};
 use tikv_util::box_err;
 
-use crate::RegionCacheMemoryEngine;
+use crate::RangeCacheMemoryEngine;
 
-/// RegionCacheWriteBatch maintains its own in-memory buffer.
+/// RangeCaheWriteBatch maintains its own in-memory buffer.
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/write_batch.rs:7]:
```suggestion
/// RangeCacheWriteBatch maintains its own in-memory buffer.
```

File: components/test_raftstore/src/node.rs [hunk 6]
```diff
@@ -526,7 +526,7 @@ pub fn new_node_cluster(id: u64, count: usize) -> Cluster<RocksEngine, NodeClust
 }
 
 // the hybrid engine with disk engine "RocksEngine" and region cache engine
-// "RegionCacheMemoryEngine" is used in the node cluster.
+// "RangeCaheMemoryEngine" is used in the node cluster.
```
Comment (afeinberg) [inline components/test_raftstore/src/node.rs:529]:
```suggestion
// "RangeCacheMemoryEngine" is used in the node cluster.
```

File: components/engine_traits/src/memory_engine.rs [hunk 7]
```diff
@@ -1,19 +1,110 @@
...
+pub trait RangeCacheEngine:
     WriteBatchExt + Iterable + Debug + Clone + Unpin + Send + Sync + 'static
 {
     type Snapshot: Snapshot;
 
-    // If None is returned, the RegionCacheEngine is currently not readable for this
+    // If None is returned, the RangeCaheEngine is currently not readable for this
     // region or read_ts.
-    // Sequence number is shared between RegionCacheEngine and disk KvEnigne to
+    // Sequence number is shared between RangeCaheEngine and disk KvEnigne to
```
Comment (afeinberg) [inline components/engine_traits/src/memory_engine.rs:19]:
```suggestion
    // Sequence number is shared between RangeCacheEngine and disk KvEngine to
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 8]
```diff
@@ -97,95 +133,98 @@ impl SnapshotList {
...
+        &self.range_manager
+    }
+
+    pub fn mut_range_manager(&mut self) -> &mut RangeManager {
+        &mut self.range_manager
     }
 }
 
-/// The RegionCacheMemoryEngine serves as a region cache, storing hot regions in
+/// The RangeCaheMemoryEngine serves as a range cache, storing hot ranges in
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:168]:
```suggestion
/// The RangeCacheMemoryEngine serves as a range cache, storing hot ranges in
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 9]
```diff
@@ -97,95 +133,98 @@ impl SnapshotList {
...
 /// the leaders' store. Incoming writes that are written to disk engine (now,
-/// RocksDB) are also written to the RegionCacheMemoryEngine, leading to a
-/// mirrored data set in the cached regions with the disk engine.
+/// RocksDB) are also written to the RangeCaheMemoryEngine, leading to a
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:170]:
```suggestion
/// RocksDB) are also written to the RangeCacheMemoryEngine, leading to a
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 10]
```diff
@@ -97,95 +133,98 @@ impl SnapshotList {
...
+/// mirrored data set in the cached ranges with the disk engine.
 ///
-/// A load/evict unit manages the memory, deciding which regions should be
-/// evicted when the memory used by the RegionCacheMemoryEngine reaches a
-/// certain limit, and determining which regions should be loaded when there is
+/// A load/evict unit manages the memory, deciding which ranges should be
+/// evicted when the memory used by the RangeCaheMemoryEngine reaches a
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:174]:
```suggestion
/// evicted when the memory used by the RangeCacheMemoryEngine reaches a
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 11]
```diff
@@ -97,95 +133,98 @@ impl SnapshotList {
...
+/// certain limit, and determining which ranges should be loaded when there is
 /// spare memory capacity.
 ///
-/// The safe point lifetime differs between RegionCacheMemoryEngine and the disk
-/// engine, often being much shorter in RegionCacheMemoryEngine. This means that
-/// RegionCacheMemoryEngine may filter out some keys that still exist in the
+/// The safe point lifetime differs between RangeCaheMemoryEngine and the disk
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:178]:
```suggestion
/// The safe point lifetime differs between RangeCacheMemoryEngine and the disk
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 12]
```diff
@@ -97,95 +133,98 @@ impl SnapshotList {
...
+/// engine, often being much shorter in RangeCaheMemoryEngine. This means that
+/// RangeCaheMemoryEngine may filter out some keys that still exist in the
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:180]:
```suggestion
/// RangeCacheMemoryEngine may filter out some keys that still exist in the
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 13]
```diff
@@ -97,95 +133,98 @@ impl SnapshotList {
...
+/// A load/evict unit manages the memory, deciding which ranges should be
+/// evicted when the memory used by the RangeCaheMemoryEngine reaches a
+/// certain limit, and determining which ranges should be loaded when there is
 /// spare memory capacity.
 ///
-/// The safe point lifetime differs between RegionCacheMemoryEngine and the disk
-/// engine, often being much shorter in RegionCacheMemoryEngine. This means that
-/// RegionCacheMemoryEngine may filter out some keys that still exist in the
+/// The safe point lifetime differs between RangeCaheMemoryEngine and the disk
+/// engine, often being much shorter in RangeCaheMemoryEngine. This means that
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:179]:
```suggestion
/// engine, often being much shorter in RangeCacheMemoryEngine. This means that
```

File: components/region_cache_memory_engine/src/engine.rs (hunk 12 above)
Comment (afeinberg) [reply components/region_cache_memory_engine/src/engine.rs:180]: More instances of that below. Looks like a find/replace had a typo. Will let you fix the rest.

File: components/region_cache_memory_engine/src/engine.rs [hunk 14]
```diff
@@ -585,38 +622,38 @@ impl Peekable for RegionCacheSnapshot {
                 user_key,
                 v_type: ValueType::Value,
                 ..
-            } if user_key == key => Ok(Some(RegionCacheDbVector(iter.value().clone()))),
+            } if user_key == key => Ok(Some(RangeCaheDbVector(iter.value().clone()))),
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:625]:
```suggestion
            } if user_key == key => Ok(Some(RangeCacheDbVector(iter.value().clone()))),
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 15]
```diff
@@ -585,38 +622,38 @@ impl Peekable for RegionCacheSnapshot {
...
-impl SnapshotMiscExt for RegionCacheSnapshot {
+impl SnapshotMiscExt for RangeCacheSnapshot {
     fn sequence_number(&self) -> u64 {
         self.sequence_number
     }
 }
 
 #[derive(Debug)]
-pub struct RegionCacheDbVector(Bytes);
+pub struct RangeCaheDbVector(Bytes);
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:644]:
```suggestion
pub struct RangeCahceDbVector(Bytes);
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 16]
```diff
@@ -585,38 +622,38 @@ impl Peekable for RegionCacheSnapshot {
...
 
-impl Deref for RegionCacheDbVector {
+impl Deref for RangeCaheDbVector {
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:646]:
```suggestion
impl Deref for RangeCacheDbVector {
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 17]
```diff
@@ -585,38 +622,38 @@ impl Peekable for RegionCacheSnapshot {
...
     type Target = [u8];
 
     fn deref(&self) -> &[u8] {
         self.0.as_slice()
     }
 }
 
-impl DbVector for RegionCacheDbVector {}
+impl DbVector for RangeCaheDbVector {}
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:654]:
```suggestion
impl DbVector for RangeCacheDbVector {}
```

File: components/region_cache_memory_engine/src/engine.rs [hunk 18]
```diff
@@ -585,38 +622,38 @@ impl Peekable for RegionCacheSnapshot {
...
 
-impl<'a> PartialEq<&'a [u8]> for RegionCacheDbVector {
+impl<'a> PartialEq<&'a [u8]> for RangeCaheDbVector {
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:656]:
```suggestion
impl<'a> PartialEq<&'a [u8]> for RangeCacheDbVector {
```

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+    }
+}
+
+#[derive(Default)]
+pub struct RangeManager {
+    // the range reflects the range of data that is accessable.
+    ranges: BTreeMap<CacheRange, RangeMeta>,
+}
+
+impl RangeManager {
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/range_manager.rs:51]: Just to make sure we're on the same page, I would use `overlaps_with_range` to find the right region cache in the write path?

**Review by afeinberg — CHANGES_REQUESTED** [review]
Looks mostly good, suggested corrections for some typos. Would like @tonyxuqqi 's review in addition to mine. Beyond that, I would suggest some unit test that test corner conditions in ranges such as from start of key range, to end of key ranges.

File: components/engine_traits/src/memory_engine.rs [hunk 20]
```diff
@@ -1,19 +1,110 @@
...
+
+    pub fn overlaps(&self, other: &CacheRange) -> bool {
+        self.start < other.end && other.start < self.end
+    }
+
+    pub fn is_sibling(&self, other: &CacheRange) -> bool {
+        self.start == other.end || self.end == other.start
+    }
+
+    pub fn split_off(&self, key: &CacheRange) -> (CacheRange, CacheRange) {
```
Comment (tonyxuqqi) [inline components/engine_traits/src/memory_engine.rs:85]: Why the `key` is CacheRange instead of a single key?

File: components/engine_traits/src/memory_engine.rs [hunk 21]
```diff
@@ -1,19 +1,110 @@
...
+            CacheRange {
+                start: key.end.clone(),
+                end: self.end.clone(),
+            },
+        )
+    }
+
+    // r1 and r2 should be unoverlap
+    pub fn merge(r1: CacheRange, r2: CacheRange) -> CacheRange {
+        assert!(!r1.overlaps(&r2));
```
Comment (tonyxuqqi) [inline components/engine_traits/src/memory_engine.rs:100]: assert!(!r1.overlaps(&r2)); 
is not necessary as it's asserted by is_sibling

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+use std::collections::{BTreeMap, BTreeSet};
+
+use engine_traits::CacheRange;
+
+use crate::engine::SnapshotList;
+
+#[derive(Debug, Default)]
+pub struct RangeMeta {
+    range_snapshot_list: BTreeMap<CacheRange, SnapshotList>,
+    ranges_evcited: BTreeSet<CacheRange>,
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:12]: When can entries in `ranges_evcited` or `ranges_unreadable` be deleted? Otherwise, these two itself would consume too much memory

File: components/region_cache_memory_engine/src/range_manager.rs (hunk 19 above)
Comment (tonyxuqqi) [reply components/region_cache_memory_engine/src/range_manager.rs:51]: should it be contains, not overlaps_with_range.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+        } else {
+            self.ranges_unreadable.insert(range.clone());
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct RangeManager {
+    // the range reflects the range of data that is accessable.
+    ranges: BTreeMap<CacheRange, RangeMeta>,
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:48]: I think to make the logic simpler and cheaper in RangeMeta,  RangeMeta is for just single range ---each entry in BTreeMap is the range that caches the data of the whole range. 
For can read status, it's per range. So we don't worry about sub-range's readability. 

And whatever region merge or split, it does not impact the logic of RangeManager. 
That is assuming original [a, c) is cached and then region split into [a, b) and [b, c), in RangeManager it's still [a, c). And both [a, b) and [b, c) should use the cache for read and write. 
And when region [b, c) is destroyed or evicted, then RangeManager would just delete [a, c) and add [a, b).

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+                return false;
+            }
+            meta.safe_ts = safe_ts;
+            true
+        } else {
+            false
+        }
+    }
+
+    pub(crate) fn overlap_with_range(&self, range: &CacheRange) -> bool {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:105]: Should we add contains method to indicate range is completely covered by a cached range?

Comment (tonyxuqqi) [reply components/region_cache_memory_engine/src/range_manager.rs:105]: In fact, we should evict the range (or add the evict hint) if it just partially covers the input range.

File: components/region_cache_memory_engine/src/engine.rs [hunk 25]
```diff
@@ -475,61 +518,55 @@ impl Iterator for RegionCacheIterator {
 }
 
 #[derive(Clone, Debug)]
-pub struct RegionCacheSnapshot {
-    region_id: u64,
+pub struct RangeCacheSnapshot {
+    range: CacheRange,
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:522]: maybe we can wrap range, snapshot_ts, sequence_number into RangeCacheSnapshotMeta struct. 
And use the RangeCacheSnapshotMeta for snapshot management.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+
+        let meta = self.ranges.get_mut(&range_key.clone()).unwrap();
+        meta.range_snapshot_list
+            .entry(range.clone())
+            .or_default()
+            .new_snapshot(read_ts);
+        true
+    }
+
+    pub(crate) fn remove_range_snapshot(&mut self, range: &CacheRange, read_ts: u64) {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:136]: pub(crate) fn remove_range_snapshot(&mut self, range: &CacheRange, read_ts: u64) {
```suggestion
    pub(crate) fn remove_range_snapshot(&mut self,  snapshot &RangeSnapshotMeta) {
```

File: components/engine_traits/src/memory_engine.rs [hunk 27]
```diff
@@ -1,19 +1,110 @@
...
+            },
+            CacheRange {
+                start: key.end.clone(),
+                end: self.end.clone(),
+            },
+        )
+    }
+
+    // r1 and r2 should be unoverlap
+    pub fn merge(r1: CacheRange, r2: CacheRange) -> CacheRange {
```
Comment (tonyxuqqi) [inline components/engine_traits/src/memory_engine.rs:99]: When is this API used?

File: components/engine_traits/src/memory_engine.rs (hunk 20 above)
Comment (SpadeA-Tang) [reply components/engine_traits/src/memory_engine.rs:85]: Because we are evicting a range.

Comment (SpadeA-Tang) [reply components/engine_traits/src/memory_engine.rs:85]: Say, now we have a range k01 - k10, and we are to evict k03-k05. After the eviction, there should be two ranges: k01-k03, and k05-k10.

File: components/engine_traits/src/memory_engine.rs (hunk 21 above)
Comment (SpadeA-Tang) [reply components/engine_traits/src/memory_engine.rs:100]: Good catch.

File: components/engine_traits/src/memory_engine.rs (hunk 27 above)
Comment (SpadeA-Tang) [reply components/engine_traits/src/memory_engine.rs:99]: It will not be called by raftstore. It will only be called by Rangemanager. Say, we now have k1-k3, and k5-k10 ranges. And we create a new range, k3-k5. We need merge to make it k1-k10 as a whole.

File: components/region_cache_memory_engine/src/range_manager.rs (hunk 24 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:105]: This is now used for assertion purpose where assert we are not create a range overlapping with existing ranges.

File: components/region_cache_memory_engine/src/engine.rs [hunk 28]
```diff
@@ -97,95 +161,101 @@ impl SnapshotList {
...
+        }
     }
-}
 
-impl Debug for RegionCacheMemoryEngine {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Region Cache Memory Engine")
+    pub fn new_range(&self, range: CacheRange) {
+        let mut core = self.core.lock().unwrap();
+        assert!(!core.range_manager.overlap_with_range(&range));
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:230]: maybe the assert can be moved inside range_manager.new_range

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 29]
```diff
@@ -0,0 +1,260 @@
...
+
+use engine_traits::CacheRange;
+
+use crate::engine::{RagneCacheSnapshotMeta, SnapshotList};
+
+#[derive(Debug, Default)]
+pub struct RangeMeta {
+    id: u64,
+    range_snapshot_list: SnapshotList,
+    evicted: bool,
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:13]: Is `evicted` needed after all? When a range is evicted, it can be moved to historical_ranges.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 30]
```diff
@@ -0,0 +1,260 @@
...
+
+    pub(crate) fn new_range(&mut self, range: CacheRange) {
+        let range_meta = RangeMeta::new(self.id_allocator.allocate_id());
+        self.ranges.insert(range, range_meta);
+    }
+
+    pub fn set_range_readable(&mut self, range: &CacheRange, set_readable: bool) {
+        if let Some(meta) = self.ranges.get_mut(range) {
+            meta.can_read = true;
+        }
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:91]: Do we expect the range must be found?

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 31]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+impl RangeMeta {
+    fn new(id: u64) -> Self {
+        Self {
+            id,
+            range_snapshot_list: SnapshotList::default(),
+            evicted: false,
+            can_read: false,
+            safe_ts: 0,
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:25]: we probably need to change the `safe_ts` to `safe_point` to be consistent with current tikv terms. 
Because today we use `safe_point` as the max timestamp to gc in compaction filter and use `safe_ts` as the min timestamp of all data are applied.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 32]
```diff
@@ -0,0 +1,260 @@
...
+
+    // Acquire a snapshot of the `range` with `read_ts`. If the range is not
+    // accessable, None will be returned. Otherwise, the range id will be returned.
+    pub(crate) fn range_snapshot(&mut self, range: &CacheRange, read_ts: u64) -> Option<u64> {
+        let Some(range_key) = self.ranges.keys().find(|&r| r.contains(range)).cloned() else {
+            return None;
+        };
+        let meta = self.ranges.get_mut(&range_key).unwrap();
+
+        if read_ts <= meta.safe_ts || meta.evicted || !meta.can_read {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:118]: I believe meta.evicted is likely not needed.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 33]
```diff
@@ -0,0 +1,260 @@
...
+    // So, here, we return a vector of ranges to denote the ranges that are ready to
+    // be removed.
+    pub(crate) fn remove_range_snapshot(
+        &mut self,
+        snapshot_meta: &RagneCacheSnapshotMeta,
+    ) -> Vec<CacheRange> {
+        if let Some(range_key) = self
+            .historical_ranges
+            .iter()
+            .find(|&(range, meta)| {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:139]: we can use range id for search in the historical_ranges

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 34]
```diff
@@ -0,0 +1,260 @@
...
+        meta.range_snapshot_list.new_snapshot(read_ts);
+        Some(meta.id)
+    }
+
+    // If the snapshot is the last one in the snapshot list of one cache range in
+    // historical_ranges, it means one or some evicted_ranges may be ready to be
+    // removed physically.
+    // So, here, we return a vector of ranges to denote the ranges that are ready to
+    // be removed.
+    pub(crate) fn remove_range_snapshot(
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:132]: Since remove_range_snapshot is called typically by the unified read pool thread, which is latency critical. 
We should not return a range list to be deleted here as it's not good to block user thread for deleting the data.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 35]
```diff
@@ -0,0 +1,260 @@
...
+            .map(|(r, _)| r.clone())
+        {
+            let meta = self.historical_ranges.get_mut(&range_key).unwrap();
+            meta.range_snapshot_list
+                .remove_snapshot(snapshot_meta.snapshot_ts);
+            if meta.range_snapshot_list.is_empty() {
+                self.historical_ranges.remove(&range_key);
+            }
+
+            return self
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:151]: I think this part should be called by GC thread.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 36]
```diff
@@ -0,0 +1,260 @@
...
+            .map(|(r, _)| r.clone())
+            .unwrap();
+        let meta = self.ranges.get_mut(&range_key).unwrap();
+        meta.range_snapshot_list
+            .remove_snapshot(snapshot_meta.snapshot_ts);
+        vec![]
+    }
+
+    // return whether the range can be already removed
+    pub(crate) fn evict_range(&mut self, evict_range: &CacheRange) -> bool {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:180]: This function needs a lock to protect

File: components/region_cache_memory_engine/src/engine.rs [hunk 37]
```diff
@@ -475,61 +549,72 @@ impl Iterator for RegionCacheIterator {
...
+impl Drop for RangeCacheSnapshot {
     fn drop(&mut self) {
         let mut core = self.engine.core.lock().unwrap();
-        let meta = core.region_metas.get_mut(&self.region_id).unwrap();
-        meta.snapshot_list.remove_snapshot(self.snapshot_ts);
+        for range_removable in core
+            .range_manager
+            .remove_range_snapshot(&self.snapshot_meta)
+        {
+            core.engine.delete_range(&self.snapshot_meta.range);
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:606]: I will optimize it in the following PR where a task will be scheduled to GC worker.

File: components/region_cache_memory_engine/src/range_manager.rs (hunk 29 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:13]: yes, no needed.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 38]
```diff
@@ -0,0 +1,260 @@
...
+// eviction range is k3-k5. k1-k10 will be splitted to three ranges: k1-k3,
+// k3-k5, and k5-k10.
+// k1-k3 and k5-k10 will be new ranges inserted in self.ranges with meta dervied
+// from meta of k1-k10 (only safe_ts and can_read will be derived). k1-k10 will
+// be removed from self.ranges and inserted to self.historical_ranges. Then,
+// k3-k5 will be in the self.evicted_ranges. Now, we cannot remove the data of
+// k3-k5 as there may be some snapshot of k1-k10. After these snapshot are
+// dropped, k3-k5 can be acutally removed.
+#[derive(Default)]
+pub struct RangeManager {
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/range_manager.rs:67]: Add a lock here to guard mutation. SInce I expect this to be read more than written to (e.g., read on any writes), consider using https://rustasync.github.io/runtime/crossbeam/sync/struct.ShardedLock.html

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 39]
```diff
@@ -0,0 +1,260 @@
...
+        }
+    }
+
+    pub(crate) fn range_snapshot_list(&self) -> &SnapshotList {
+        &self.range_snapshot_list
+    }
+}
+
+#[derive(Default)]
+struct IdAllocator(u64);
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/range_manager.rs:45]: Either document that this should be called under a lock, or use an atomic.

File: components/region_cache_memory_engine/src/range_manager.rs [hunk 40]
```diff
@@ -0,0 +1,260 @@
...
+                return false;
+            }
+            meta.safe_ts = safe_ts;
+            true
+        } else {
+            false
+        }
+    }
+
+    pub(crate) fn overlap_with_range(&self, range: &CacheRange) -> bool {
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/range_manager.rs:106]: Can we also add `contains(&self, key: &[u8])` or `get_range_meta_for_key(self, key; &[u8]) -> Option<RangeMeta>`?

File: components/region_cache_memory_engine/src/engine.rs [hunk 41]
```diff
@@ -32,63 +37,122 @@ fn cf_to_id(cf: &str) -> usize {
...
+        assert!(self.recorder.lock().unwrap().is_empty());
+    }
+}
+
+/// A single global set of skiplists shared by all cached ranges
 #[derive(Clone)]
-pub struct RegionMemoryEngine {
-    data: [Arc<Skiplist<InternalKeyComparator>>; 3],
+pub struct SkiplistEngine {
+    data: [Arc<Skiplist<InternalKeyComparator, GlobalMemoryLimiter>>; 3],
```
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:84]: i may add it myself, but may be (pub) struct SkiplistEngine.

File: components/region_cache_memory_engine/src/range_manager.rs (hunk 38 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:67]: It's a field of RangeCacheMemoryEngineCore which is protected by a mutex in RangeCacheMemoryEngine.

File: components/region_cache_memory_engine/src/range_manager.rs (hunk 39 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:45]: The whole RangeManager is protected by a lock.

File: components/region_cache_memory_engine/src/range_manager.rs (hunk 40 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:106]: what's the use cases?

**Review by afeinberg — COMMENTED** [review]
Looks mostly good, but please add a "contains_key" or "contains" method to check if a key is cached to RangeManager.

File: components/region_cache_memory_engine/src/range_manager.rs (hunk 40 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:106]: Ok, get it.

**Review by tonyxuqqi — APPROVED** [review]

**Review by afeinberg — APPROVED** [review]
LGTM. Note: you'll need to update WriteBatch (my changes just went in), but should be fairly simple changes.

**Review by bufferflies — APPROVED** [review]

File: components/engine_traits/src/memory_engine.rs [hunk 42]
```diff
@@ -1,19 +1,105 @@
...
+        Self {
+            start: enc_start_key(region),
+            end: enc_end_key(region),
+        }
+    }
+}
+
+impl PartialOrd for CacheRange {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        if self.end <= other.start {
```
Comment (overvenus) [inline components/engine_traits/src/memory_engine.rs:45]: What if `self.end` is empty? An empty end_key in region means a infinite large key.

Comment (SpadeA-Tang) [reply components/engine_traits/src/memory_engine.rs:45]: CacheRange has data_key prefix
